src/compiler/builder.ts

```diff
@@ -409,31 +406,72 @@ namespace ts {
+     *  Similar to cleanSemanticDiagnosticsOfAffectedFile
      */
-    function cleanSemanticDiagnosticsOfAffectedFile(state: BuilderProgramState, affectedFile: SourceFile) {
-        if (removeSemanticDiagnosticsOf(state, affectedFile.path)) {
-            // If there are no more diagnostics from old cache, done
+    function handleDtsMayChangeOfAffectedFile(state: BuilderProgramState, affectedFile: SourceFile, cancellationToken: CancellationToken | undefined, computeHash: BuilderState.ComputeHash) {
+        removeSemanticDiagnosticsOf(state, affectedFile.path);
+
+        // If affected files is everything except default librarry, then nothing more to do
+        if (state.allFilesExcludingDefaultLibraryFile === state.affectedFiles) {
+            if (!state.cleanedDiagnosticsOfLibFiles) {
+                state.cleanedDiagnosticsOfLibFiles = true;
+                const program = Debug.assertDefined(state.program);
+                const options = program.getCompilerOptions();
+                forEach(program.getSourceFiles(), f =>
+                    program.isSourceFileDefaultLibrary(f) &&
+                    !skipTypeChecking(f, options) &&
+                    removeSemanticDiagnosticsOf(state, f.path)
+                );
+            }
             return;
         }
 
-        // Clean lib file diagnostics if its all files excluding default files to emit
-        if (state.allFilesExcludingDefaultLibraryFile === state.affectedFiles && !state.cleanedDiagnosticsOfLibFiles) {
-            state.cleanedDiagnosticsOfLibFiles = true;
+        forEachReferencingModulesOfExportOfAffectedFile(state, affectedFile, (state, path) => handleDtsMayChangeOf(state, path, cancellationToken, computeHash));
+    }
+
+    /**
+     * Handle the dts may change, so they need to be added to pending emit if dts emit is enabled,
+     * Also we need to make sure signature is updated for these files
+     */
+    function handleDtsMayChangeOf(state: BuilderProgramState, path: Path, cancellationToken: CancellationToken | undefined, computeHash: BuilderState.ComputeHash) {
+        removeSemanticDiagnosticsOf(state, path);
+
+        if (!state.changedFilesSet.has(path)) {
             const program = Debug.assertDefined(state.program);
-            const options = program.getCompilerOptions();
-            if (forEach(program.getSourceFiles(), f =>
-                program.isSourceFileDefaultLibrary(f) &&
-                !skipTypeChecking(f, options) &&
-                removeSemanticDiagnosticsOf(state, f.path)
-            )) {
-                return;
+            const sourceFile = program.getSourceFileByPath(path);
+            if (sourceFile) {
+                BuilderState.updateShapeSignature(state, program, sourceFile, Debug.assertDefined(state.currentAffectedFilesSignatures), cancellationToken, computeHash, state.currentAffectedFilesExportedModulesMap);

```

Comment:
nitpick: this line is too long (for github, and my 1440px-wide screen)

Reply:
To aid my understanding of the system: is updateShapeSignature required for getEmitDeclarations to return the right answer? It seems like addtoAffectedFilesPendingEmit is the central part of the fix, but I don't understand how updateShapeSignature is needed, especially since its comment implies that it's pure. It looks like it's called for side effects here.

Reply:
`updateShapeSignature ` is required to ensure that we have right signature for files pending emit (these don't need any js file emit but do need d.ts emit or need to update diagnostics.) Since the d.ts. output is signature for the file that we have already handled we need to update the signature.

Reply:
Adding comment here for clarity